File: apps/browser-extension/src/entrypoints/popup/utils/ConversionUtility.ts
/**
 * Utility class for conversion operations.
 * TODO: make this a shared utility class in root /shared/ folder so we can reuse it between browser extension/mobile app
 * and possibly WASM client.
 */
class ConversionUtility {
  /**
   * Convert all anchor tags to open in a new tab.
   * @param html HTML input.
   * @returns HTML with all anchor tags converted to open in a new tab when clicked on.
   *
   * Note: same implementation exists in c-sharp version in AliasVault.Shared.Utilities.ConversionUtility.cs
   */
  public convertAnchorTagsToOpenInNewTab(html: string): string {
    try {
      // Create a DOM parser
      const parser = new DOMParser();
      const doc = parser.parseFromString(html, 'text/html');

      // Select all anchor tags with href attribute
      const anchors = doc.querySelectorAll('a[href]');

      if (anchors.length > 0) {
        anchors.forEach((anchor: Element) => {
          // Handle target attribute
          if (!anchor.hasAttribute('target') || anchor.getAttribute('target') !== '_blank') {
            anchor.setAttribute('target', '_blank');
          }

          // Handle rel attribute for security
          if (!anchor.hasAttribute('rel')) {
            anchor.setAttribute('rel', 'noopener noreferrer');
          } else {
            const relValue = anchor.getAttribute('rel') ?? '';
            const relValues = new Set(relValue.split(' ').filter(val => val.trim() !== ''));

            relValues.add('noopener');
            relValues.add('noreferrer');

            anchor.setAttribute('rel', Array.from(relValues).join(' '));
          }
        });
      }

      return doc.documentElement.outerHTML;
    } catch (ex) {
      // Log the exception
      console.error(`Error in convertAnchorTagsToOpenInNewTab: ${ex instanceof Error ? ex.message : String(ex)}`);

      // Return the original HTML if an error occurs
      return html;
    }
  }

  /**
   * Normalize a username by converting it to lowercase and trimming whitespace.
   * @param username The username to normalize.
   * @returns The normalized username.
   */
  public normalizeUsername(username: string): string {
    return username.toLowerCase().trim();
  }
}

export default new ConversionUtility();

File: apps/server/Shared/AliasVault.Shared/Utilities/ConversionUtility.cs
//-----------------------------------------------------------------------
// <copyright file="ConversionUtility.cs" company="aliasvault">
// Copyright (c) aliasvault. All rights reserved.
// Licensed under the AGPLv3 license. See LICENSE.md file in the project root for full license information.
// </copyright>
//-----------------------------------------------------------------------

namespace AliasVault.Shared.Utilities;

using HtmlAgilityPack;

/// <summary>
/// Class which contains various helper methods for data conversion.
/// </summary>
public static class ConversionUtility
{
    /// <summary>
    /// Convert all anchor tags to open in a new tab.
    /// </summary>
    /// <param name="html">HTML input.</param>
    /// <returns>HTML with all anchor tags converted to open in a new tab when clicked on.</returns>
    /// <remarks>
    /// Note: same implementation exists in browser extension Typescript version in ConversionUtility.ts.
    /// </remarks>
    public static string ConvertAnchorTagsToOpenInNewTab(string html)
    {
        try
        {
            var doc = new HtmlDocument();
            doc.LoadHtml(html);

            var anchors = doc.DocumentNode.SelectNodes("//a[@href]");
            if (anchors != null)
            {
                foreach (var anchor in anchors)
                {
                    var targetAttr = anchor.Attributes["target"];
                    if (targetAttr == null)
                    {
                        anchor.SetAttributeValue("target", "_blank");
                    }
                    else if (targetAttr.Value != "_blank")
                    {
                        targetAttr.Value = "_blank";
                    }

                    // Add rel="noopener noreferrer" for security
                    var relAttr = anchor.Attributes["rel"];
                    if (relAttr == null)
                    {
                        anchor.SetAttributeValue("rel", "noopener noreferrer");
                    }
                    else if (!relAttr.Value.Contains("noopener") || !relAttr.Value.Contains("noreferrer"))
                    {
                        var relValues = new HashSet<string>(relAttr.Value.Split(' ', StringSplitOptions.RemoveEmptyEntries));
                        relValues.Add("noopener");
                        relValues.Add("noreferrer");
                        anchor.SetAttributeValue("rel", string.Join(" ", relValues));
                    }
                }
            }

            return doc.DocumentNode.OuterHtml;
        }
        catch (Exception ex)
        {
            // Log the exception
            Console.WriteLine($"Error in ConvertAnchorTagsToOpenInNewTab: {ex.Message}");

            // Return the original HTML if an error occurs
            return html;
        }
    }
}
